File: apps/backend/internal/domain/model/canvas.go
package model

import (
	"time"

	"gorm.io/gorm"
)

type Canvas struct {
	gorm.Model
	OwnerID      uint         `gorm:"not null" json:"owner_id"`
	OrgID        *uint        `gorm:"index" json:"org_id,omitempty"`
	Owner        User         `json:"owner,omitempty"`
	Name         string       `gorm:"not null" json:"name"`
	Description  string       `gorm:"size:512" json:"description,omitempty"`
	CanvasType   string       `gorm:"default:'inspiration'" json:"canvas_type"` // inspiration|workflow
	ProjectID    *uint        `json:"project_id,omitempty"`
	Stage        string       `json:"stage"`    // script_analysis|asset_prep|storyboard|generation|editing
	RefType      string       `json:"ref_type"` // script|setting|asset_slot
	RefID        *uint        `json:"ref_id,omitempty"`
	Visibility   string       `gorm:"default:'private';index" json:"visibility"` // private|public
	WorkflowKey  string       `gorm:"size:160;index" json:"workflow_key,omitempty"`
	WorkflowTags string       `gorm:"type:text" json:"workflow_tags,omitempty"` // JSON array for marketplace filtering
	PublishedAt  *time.Time   `json:"published_at,omitempty"`
	Nodes        []CanvasNode `gorm:"foreignKey:CanvasID" json:"nodes,omitempty"`
	Edges        []CanvasEdge `gorm:"foreignKey:CanvasID" json:"edges,omitempty"`
}

// CanvasNode stores a React Flow node persisted to DB.
// Data is a JSON string: { source, resourceId?, prompt?, modelId?, status?, taskId?, error? }
// Type values:
//
//	media:    text | image | video | audio
//	tools:    canvas | ref_image_gen | ref_video_gen | multi_angle | style_transfer | motion_imitation
type CanvasNode struct {
	gorm.Model
	CanvasID uint    `gorm:"not null" json:"canvas_id"`
	NodeID   string  `gorm:"not null" json:"node_id"` // React Flow node id
	Type     string  `gorm:"not null" json:"type"`
	Label    string  `json:"label"`
	PosX     float64 `json:"pos_x"`
	PosY     float64 `json:"pos_y"`
	Data     string  `json:"data"` // JSON blob
}

type CanvasEdge struct {
	gorm.Model
	CanvasID     uint   `gorm:"not null" json:"canvas_id"`
	EdgeID       string `gorm:"not null" json:"edge_id"`
	Source       string `gorm:"not null" json:"source"`  // source node_id
	Target       string `gorm:"not null" json:"target"`  // target node_id
	SourceHandle string `json:"source_handle,omitempty"` // output port id
	TargetHandle string `json:"target_handle,omitempty"` // input port id
}

type CanvasRun struct {
	gorm.Model
	CanvasID          uint         `gorm:"not null" json:"canvas_id"`
	Status            string       `gorm:"default:'pending';index" json:"status"`    // pending | running | done | failed
	InputValues       string       `json:"input_values,omitempty"`                   // JSON object keyed by input node_id
	OutputValues      string       `gorm:"type:text" json:"output_values,omitempty"` // JSON object keyed by output node_id / param name
	Error             string       `json:"error,omitempty"`
	GraphSnapshot     string       `gorm:"type:text" json:"graph_snapshot,omitempty"`
	SnapshotHash      string       `gorm:"size:64;index" json:"snapshot_hash,omitempty"`
	SnapshotNodeCount int          `json:"snapshot_node_count"`
	SnapshotEdgeCount int          `json:"snapshot_edge_count"`
	StartedAt         *time.Time   `json:"started_at,omitempty"`
	FinishedAt        *time.Time   `json:"finished_at,omitempty"`
	Tasks             []CanvasTask `gorm:"foreignKey:CanvasRunID" json:"tasks,omitempty"`
}

type CanvasTask struct {
	gorm.Model
	CanvasNodeID   uint         `gorm:"not null" json:"canvas_node_id"`
	CanvasRunID    *uint        `json:"canvas_run_id,omitempty"`
	CanvasRun      *CanvasRun   `gorm:"foreignKey:CanvasRunID" json:"canvas_run,omitempty"`
	NodeID         string       `gorm:"index" json:"node_id,omitempty"`
	NodeLabel      string       `json:"node_label,omitempty"`
	NodeType       string       `json:"node_type,omitempty"`
	Status         string       `gorm:"default:'pending'" json:"status"` // pending | running | done | failed
	ProviderTaskID string       `json:"provider_task_id,omitempty"`
	Error          string       `json:"error,omitempty"`
	InputValues    string       `gorm:"type:text" json:"input_values,omitempty"`  // JSON: input port id -> values
	OutputValues   string       `gorm:"type:text" json:"output_values,omitempty"` // JSON: output port id -> value
	ResourceID     *uint        `json:"resource_id,omitempty"`
	Resource       *RawResource `gorm:"foreignKey:ResourceID" json:"resource,omitempty"`
}

type CanvasEntityWriteAudit struct {
	gorm.Model
	CanvasID           uint   `gorm:"index" json:"canvas_id"`
	CanvasRunID        uint   `gorm:"index" json:"canvas_run_id"`
	CanvasNodeID       string `gorm:"index" json:"canvas_node_id"`
	PortID             string `gorm:"not null;index" json:"port_id"`
	EntityKind         string `gorm:"not null;index:idx_canvas_entity_write_audit_entity" json:"entity_kind"`
	EntityID           uint   `gorm:"not null;index:idx_canvas_entity_write_audit_entity" json:"entity_id"`
	UserID             uint   `gorm:"index" json:"user_id"`
	OldValueJSON       string `gorm:"type:text" json:"old_value_json,omitempty"`
	NewValueJSON       string `gorm:"type:text" json:"new_value_json,omitempty"`
	ResourceBindingIDs string `gorm:"type:text" json:"resource_binding_ids,omitempty"`
}

// CanvasOutput records an explicit output target from a canvas run to a semantic entity or
// legacy entity. It complements the write audit with product-level intent.
type CanvasOutput struct {
	gorm.Model
	ProjectID    uint   `gorm:"not null;index" json:"project_id"`
	CanvasID     uint   `gorm:"not null;index" json:"canvas_id"`
	CanvasRunID  *uint  `gorm:"index" json:"canvas_run_id,omitempty"`
	CanvasNodeID string `gorm:"index" json:"canvas_node_id"`
	PortID       string `gorm:"not null;index" json:"port_id"`
	OwnerType    string `gorm:"not null;index:idx_canvas_output_owner" json:"owner_type"`
	OwnerID      uint   `gorm:"not null;index:idx_canvas_output_owner" json:"owner_id"`
	OutputType   string `gorm:"not null;default:'resource';index" json:"output_type"` // resource|field|candidate|note
	ResourceID   *uint  `gorm:"index" json:"resource_id,omitempty"`
	TargetField  string `json:"target_field"`
	ValueJSON    string `gorm:"type:text" json:"value_json"`
	Status       string `gorm:"not null;default:'pending';index" json:"status"` // pending|attached|applied|rejected
	MetadataJSON string `gorm:"type:text" json:"metadata_json"`
}

File: apps/desktop/src/types/canvas.ts
import type { RawResource } from './resources'

// Canvas
export type MediaNodeType = 'text' | 'image' | 'video' | 'audio'
export type ToolNodeType = 'canvas' | 'reference_to_image' | 'reference_to_video'
export type SemanticEntityKind = 'script' | 'segment' | 'scene_moment' | 'setting' | 'asset_slot' | 'content_unit'
export type SpecialNodeType = 'input' | 'output' | 'resource_sink' | 'approval' | 'text_gen' | 'ai_gen' | 'group' | 'plugin_card'
export type PluginNodeType = string & { readonly __pluginNodeType?: unique symbol }
export type NodeType = MediaNodeType | ToolNodeType | SpecialNodeType | PluginNodeType
export type NodeSource = 'upload' | 'ai' | 'manual'
export type CanvasTaskStatus = 'idle' | 'pending' | 'running' | 'done' | 'failed'
export type CanvasType = 'inspiration' | 'workflow'
export type CanvasParamType = 'text' | 'image' | 'video' | 'audio' | 'json' | 'number' | 'boolean' | 'resource'
export type CanvasResourceMediaType = 'image' | 'video' | 'audio' | 'text' | 'any'
export type CanvasResourceRole =
  | 'generic'
  | 'first_frame'
  | 'last_frame'
  | 'reference_image'
  | 'reference_video'
  | 'reference_audio'
  | 'style_reference'
  | 'target_image'
  | 'target_video'
  | 'source_audio'
  | 'speech_audio'
  | 'voice_sample'
  | 'target_voice'
  | 'transcript'
  | 'mask'
export type CanvasRunStatus = 'pending' | 'running' | 'done' | 'failed'
export type CanvasPortType = CanvasParamType

export interface CanvasNodeModelDiagnostics {
  canvas_id: number
  node_id: string
  node_label: string
  node_type: string
  capability?: string
  status: 'ok' | 'missing_model_selection' | 'route_error' | 'not_applicable' | 'invalid_node_data' | 'ai_service_unavailable' | 'missing_capability' | string
  problems?: string[]
  next_actions?: string[]
  raw_model_fields?: Record<string, unknown>
  data_model_id?: string
  executable?: boolean
  executable_model_id?: string
  available_model_count: number
  available_models?: Array<{
    model_id: string
    display_name: string
    is_default?: boolean
    capabilities?: string[]
  }>
  route?: {
    model_id: string
    provider_model_id?: string
    selection_reason?: string
  }
}

export interface CanvasPortDef {
  id: string
  aliases?: string[]
  label?: string
  labelKey?: string
  type: CanvasPortType
  mediaType?: CanvasResourceMediaType
  acceptedMediaTypes?: CanvasResourceMediaType[]
  role?: CanvasResourceRole
  order?: number
  required?: boolean
  maxCount?: number
  deprecated?: boolean
  description?: string
}

export interface EntityWorkflowField {
  readable: boolean
  writable: boolean
  portId: string
  aliases?: string[]
  required?: boolean
  maxCount?: number
}

export interface EntityWorkflowBinding {
  role: string
  slot: string
  isPrimary: boolean
  multiple: boolean
}

export interface EntitySemanticFieldIO {
  readable: boolean
  writable: boolean
  required?: boolean
  maxCount?: number
}

export interface EntitySemanticSchemaField {
  id: string
  aliases?: string[]
  deprecated?: boolean
  labelKey: string
  fallbackLabel: string
  valueType: CanvasPortType
  control: 'input' | 'textarea' | 'select' | 'number' | 'checkbox' | 'json_editor' | 'resource_picker' | 'resource_gallery' | 'related_entity_list' | 'readonly_text' | 'computed' | string
  readonly?: boolean
  layout?: {
    width?: string
    relation?: string
    nestedKind?: string
  }
  io: EntitySemanticFieldIO
  binding?: EntityWorkflowBinding
  validation?: {
    required?: boolean
    enum?: string[]
    min?: number
    max?: number
  }
}

export interface EntitySemanticSchemaSection {
  id: string
  labelKey: string
  fallbackLabel: string
  layout?: {
    variant?: string
    columns?: number
  }
  fields: EntitySemanticSchemaField[]
}

export interface EntitySemanticSchema {
  kind: SemanticEntityKind
  schemaVersion?: number
  projection?: string
  compatibility?: EntitySchemaCompatibility
  labelKey: string
  fallbackLabel: string
  layout?: {
    variant?: string
  }
  sections: EntitySemanticSchemaSection[]
}

export interface EntitySchemaCompatibility {
  currentVersion: number
  minCompatibleVersion: number
  fieldAliases?: Record<string, string[]>
  deprecatedFields?: string[]
  migrations?: EntitySchemaMigration[]
}

export interface EntitySchemaMigration {
  fromVersion: number
  toVersion: number
  kind: string
  fieldId?: string
  fromFieldId?: string
  toFieldId?: string
  description?: string
}

export interface EntitySchemaActionHint {
  kind: string
  fieldId?: string
  fromFieldId?: string
  toFieldId?: string
  description: string
}

export interface EntitySchemaMigrationReport {
  kind: SemanticEntityKind
  schemaVersion: number
  currentVersion: number
  minCompatibleVersion: number
  fieldAliases?: Record<string, string[]>
  deprecatedFields?: string[]
  migrations?: EntitySchemaMigration[]
  actions: EntitySchemaActionHint[]
}

export interface EntitySemanticValues {
  kind: SemanticEntityKind
  id: number
  schemaVersion: number
  values: Record<string, unknown>
}

export interface EntityWorkflowSchemaField {
  id: string
  aliases?: string[]
  deprecated?: boolean
  labelKey: string
  fallbackLabel: string
  valueType: CanvasPortType
  control: 'input' | 'textarea' | 'select' | 'number' | 'checkbox' | 'json_editor' | 'resource_picker' | 'resource_gallery' | 'readonly_text' | 'computed' | string
  readonly?: boolean
  layout?: {
    width?: string
    relation?: string
    nestedKind?: string
  }
  workflow: EntityWorkflowField
  binding?: EntityWorkflowBinding
  validation?: {
    required?: boolean
    enum?: string[]
    min?: number
    max?: number
  }
}

export interface EntityWorkflowSchemaSection {
  id: string
  labelKey: string
  fallbackLabel: string
  layout?: {
    variant?: string
    columns?: number
  }
  fields: EntityWorkflowSchemaField[]
}

export interface EntityWorkflowSchema {
  kind: SemanticEntityKind
  schemaVersion?: number
  projection?: string
  compatibility?: EntitySchemaCompatibility
  labelKey: string
  fallbackLabel: string
  layout?: {
    variant?: string
  }
  sections: EntityWorkflowSchemaSection[]
}

export interface CanvasPortValue {
  type: CanvasPortType
  resource_id?: number
  media_type?: CanvasResourceMediaType
  role?: CanvasResourceRole
  resource?: RawResource
  text?: string
  json?: unknown
  number?: number
  boolean?: boolean
}

export type CanvasStage = 'script_analysis' | 'asset_prep' | 'storyboard' | 'generation' | 'editing'

export type CanvasExecutableCapability = 'text_generation' | 'image_generation' | 'video_generation' | 'audio_generation'

export interface CanvasExecutableReferenceAsset {
  resource_id?: number
  media_type?: CanvasResourceMediaType
  role: CanvasResourceRole
}

export interface CanvasExecutableSpec {
  executor: 'ai_model' | 'plugin_http'
  capability: CanvasExecutableCapability
  operation?: string
  modelId?: string
  prompt?: string
  inputResourceIds?: number[]
  referenceAssets?: CanvasExecutableReferenceAsset[]
  aspectRatio?: string
  duration?: number
  params?: Record<string, unknown>
}

export interface CanvasNodeData {
  source: NodeSource
  resourceId?: number
  resource?: RawResource
  prompt?: string
  modelId?: string      // public logical model ID preferred for routing
  modelOperation?: string                               // canonical generation operation, e.g. text_to_image or first_last_frame_to_video
  referencedCanvasId?: number                            // workflow canvas used by a canvas reference node
  referencedCanvasName?: string                          // denormalized workflow name for reference card display
  inputResourceIds?: number[]                             // selected resource inputs for full tool cards
  params?: Record<string, unknown>                        // generation parameters for canvas AI nodes
  status?: CanvasTaskStatus
  taskId?: number
  error?: string
  runDiagnostics?: CanvasNodeModelDiagnostics                  // transient UI diagnostic, not persisted
  textContent?: string                                     // manual text nodes
  inputValue?: string                                      // input nodes
  paramName?: string                                       // input/output parameter name
  paramType?: CanvasParamType                              // input/output parameter type
  paramOrder?: number                                      // input/output parameter order
  approvalStatus?: 'waiting' | 'approved' | 'rejected'    // approval nodes
  // ai_gen node fields
  refNodeIds?: string[]                                    // referenced node IDs for @mentions
  outputType?: 'image' | 'video' | 'text'                 // what to auto-generate after
  // group node fields
  groupId?: string                                         // which group this node belongs to
  isGroup?: boolean                                        // true for group container nodes
  groupLabel?: string
  groupWidth?: number
  groupHeight?: number
  // display mode
  cardMode?: 'compact' | 'detail' | 'full'
  // local plugin card fields
  pluginId?: string
  pluginName?: string
  pluginVersion?: string
  pluginRuntime?: 'trusted_local' | 'backend_ai_model' | 'backend_http' | 'callback'
  pluginArgs?: Record<string, unknown>
  pluginResultText?: string
  pluginResultData?: unknown
  pluginLastRunAt?: string
  executableSpec?: CanvasExecutableSpec
  inputPorts?: CanvasPortDef[]
  outputPorts?: CanvasPortDef[]
  // injected at runtime by CanvasEditorPage (not persisted)
  canvasId?: string
  rfNodeId?: string
}

export interface CanvasNodeModel {
  ID: number
  canvas_id: number
  node_id: string
  type: NodeType
  label: string
  pos_x: number
  pos_y: number
  data: string // JSON of CanvasNodeData
}

export interface CanvasEdgeModel {
  ID: number
  canvas_id: number
  edge_id: string
  source: string
  target: string
  source_handle?: string
  target_handle?: string
}

export interface Canvas {
  ID: number
  owner_id: number
  name: string
  stage?: CanvasStage
  canvas_type?: CanvasType
  ref_type?: string
  ref_id?: number
  nodes?: CanvasNodeModel[]
  edges?: CanvasEdgeModel[]
}

export interface CanvasTask {
  ID: number
  canvas_node_id: number
  canvas_run_id?: number
  node_id?: string
  node_label?: string
  node_type?: string
  status: CanvasTaskStatus
  provider_task_id?: string
  error?: string
  input_values?: string
  output_values?: string
  resource_id?: number
  resource?: RawResource
  CreatedAt: string
}

export interface CanvasRun {
  ID: number
  canvas_id: number
  status: CanvasRunStatus
  input_values?: string
  output_values?: string
  error?: string
  graph_snapshot?: string
  snapshot_hash?: string
  snapshot_node_count?: number
  snapshot_edge_count?: number
  started_at?: string
  finished_at?: string
  tasks?: CanvasTask[]
  CreatedAt: string
  UpdatedAt: string
}

export interface CanvasEntityWriteAudit {
  ID: number
  canvas_id: number
  canvas_run_id?: number
  canvas_node_id?: string
  port_id: string
  entity_kind: SemanticEntityKind
  entity_id: number
  user_id?: number
  old_value_json?: string
  new_value_json?: string
  resource_binding_ids?: string
  CreatedAt: string
}
